### ISSUE TITLE
macOS SCEP renewals fail over HTTPS

### ISSUE THREAD

**Issue by elyscape** [smallstep/certificates#746]
### Subject of the issue
The macOS SCEP client has some strange behavior. If the SCEP endpoint is served over HTTPS and accepts client certificates, it will use the certificate that it is attempting to renew, but there are two very important caveats:

1. It will only do so if it can build a certificate chain to a root that it trusts
2. It only presents the leaf certificate to the server

Item number 1 is interesting because, despite retrieving the intermediate CA certificate as part of the SCEP enrollment process, macOS does not save it to the keychain. This seems like a bug but, as a result, administrators who create configuration profiles that include a SCEP payload but not the intermediate CA cert will find their systems not being able to actually use the certificate generated by the process.

Item number 2 is a problem for us, because `step-ca` is currently written [to verify client certificates, if presented, against only the root CA certificate][verify]. As a result, when a properly-configured macOS machine attempts to renew the certificate, `step-ca` will fail to validate the presented client certificate and drop the connection.

### Your environment
* OS - macOS 12.0.1
* Version - 0.17.6

### Steps to reproduce
1. Download and extract [this demo environment](https://github.com/smallstep/certificates/files/7543568/environment.zip) on a Mac
2. From the directory into which it has been extracted, run `step-ca config/ca.json`
3. Install the configuration profile named "Sample SCEP Profile - No Chain.mobileconfig"
    1. Open the file
    2. Launch System Preferences
    3. Click "Profiles"
    4. Click on "Sample SCEP Profile - No Chain"
    5. Click "Install"
    6. Click "Install" again
4. Renew the certificate by clicking the "Update" button in the bottom-right of the window
5. Remove the profile
6. Repeat steps 3-5 with the profile named "Sample SCEP Profile - With Chain.mobileconfig"

### Expected behaviour
Renewal succeeds with either configuration profile.

### Actual behaviour
Renewal fails with the "Sample SCEP Profile - With Chain" profile. `step-ca` logs the following:
```
2021/11/15 20:17:21 /usr/local/Cellar/go/1.17.2/libexec/src/net/http/server.go:3157: http: TLS handshake error from 127.0.0.1:62298: tls: failed to verify client certificate: x509: certificate signed by unknown authority
```

[verify]: https://github.com/smallstep/certificates/blob/v0.17.6/ca/ca.go#L420-L437

**Comment by elyscape** [comment]
This could likely be worked around by changing `root` in the config to be an array including both the root CA cert and the intermediate CA cert, but that seems suboptimal.

**Comment by elyscape** [comment]
Update: on further research, it's possible that macOS's behavior is actually a result of `step-ca` not correctly following [the SCEP spec][spec].

> If an intermediate CA is in use, a certificates-only CMS SignedData message with a certificate chain consisting of all CA certificates is returned. Otherwise, the CA certificate itself is returned.

Currently, `step-ca` responds to `GetCACert` requests with only the intermediate certificate. It should include the root certificate as well.

[spec]: https://www.rfc-editor.org/rfc/rfc8894.html#section-2.2-2

**Comment by hslatman** [comment]
Hi @elyscape,

I'll have a look at it. 

The SCEP RFCs are not very clear about all specifics. This often leads to incompatibilities when deploying it.

Currently in `step-ca` the intermediate CA key/certificate is used as the signer (of CSRs) and (en|de)crypter (of selected SCEP requests) by the SCEP provisioner, so what the SCEP client receives is in fact the issuing CA certificate. The RFCs also mention an RA, which is an intermediate party between the client and the issuing CA, which can have its own key/certificate. `step-ca` currently doesn't support that yet. Your SCEP client may be hitting an issue as described [here](https://github.com/openxpki/openxpki/issues/73), which is part of what I left as a TODO at https://github.com/smallstep/certificates/blob/master/scep/authority.go#L142-L143. Doing this completely right will take some time, as SCEP client implementations can be quite specific.

Regarding the second issue: SCEP + HTTPS can lead to interoperability problems, as described in [7.10. Use of HTTP](https://www.rfc-editor.org/rfc/rfc8894.html#name-use-of-http). Your case _may_ be an example of this, but it's hard to tell without actually trying and seeing what your SCEP client is doing and not doing, specifically. What is the SCEP client implementation you're using? Is it part of an MDM solution for macOS?

Can you try requesting the certificate using HTTP instead and see if that does work with your client? You'll need `insecureAddress` to be configured in your `ca.json` for that.

**Comment by elyscape** [comment]
> Your SCEP client may be hitting an issue as described here, which is part of what I left as a TODO at https://github.com/smallstep/certificates/blob/master/scep/authority.go#L142-L143.

I believe that is what is happening.

> What is the SCEP client implementation you're using? Is it part of an MDM solution for macOS?

It's the macOS built-in SCEP client, which is managed through configuration profiles. I built a test environment and configuration profiles that demonstrate the issue and linked them in the "Steps to reproduce" section.

> Can you try requesting the certificate using HTTP instead and see if that does work with your client?

This works around the issue, as does putting a reverse proxy in front of `step-ca`, but these do so by effectively stripping the client certificate from the connection. Ideally that would not be necessary.

**Comment by elyscape** [comment]
> Your SCEP client may be hitting an issue as described [here](https://github.com/openxpki/openxpki/issues/73)

As an aside, it's worth noting the commit message from their fix for the issue in https://github.com/openxpki/openxpki/commit/c5901fe4e79231d27e5a48c79398add62b6a361a:

> The standard is a bit vague but as most clients seem to expect the root cert in the getcacert response we now include it by default. For enhanced security, you can strip it from the chain setting a config flag.

As an additional data point, Active Directory Certificate Services returns the full chain in response to a `GetCACert` request.

**Comment by elyscape** [comment]
> On the other hand, it may be possible to change `step-ca` to not request client certificates when a SCEP endpoint is hit, as it's something that is not part of the SCEP spec.

This is probably not possible, as client certificates are requested during TLS session establishment and the SCEP endpoint is only distinguishable once an HTTP request has been made.

**Comment by elyscape** [comment]
I made a test build of `step-ca` that returns the full cert chain in response to `GetCACert` requests and found the following:

1. macOS still doesn't save the CA certs provided in response, and as a result the client cert isn't actually usable by applications (or the system) unless the chain is installed to the keychain independently of the SCEP profile. This seems like a bug in macOS.
2. If `GetCACert` responds with a certificate chain, the intermediate CA cert's `keyUsage` extension _must_ include `digitalSignature` and `keyEncipherment`, or else the SCEP client will abort the request.

Item 2 is in line with [the spec](https://www.rfc-editor.org/rfc/rfc8894.html#name-certificate-authority):

> Since the client is expected to perform signature verification and optionally encryption using the CA certificate, the keyUsage extension in the CA certificate MUST indicate that it is valid for digitalSignature and keyEncipherment (if the key is to be used for en/decryption) alongside the usual CA usages of keyCertSign and/or cRLSign.

It's curious that the macOS SCEP client doesn't require this to be the case if only a single CA cert is returned by `GetCACert`. That seems like it might be another bug in macOS. Regardless, this is a potential issue for us because `scep-ca`'s `intermediate-ca` profile only includes `certSign` and `crlSign` in its `keyUsage` extension, so that's a fun wrinkle.

**Comment by elyscape** [comment]
I am attaching [an environment](https://github.com/smallstep/certificates/files/7551372/fixed-usage.zip) similar to that posted in the original post, but with an intermediate CA cert that also has the `digitalSignature` and `keyEncipherment` key usages.

**Comment by hslatman** [comment]
> It's the macOS built-in SCEP client, which is managed through configuration profiles. I built a test environment and configuration profiles that demonstrate the issue and linked them in the "Steps to reproduce" section.

I'll have a look at them. Will need to find out what binary is responsible for it and see what I can find out about it.

> This works around the issue, as does putting a reverse proxy in front of step-ca, but these do so by effectively stripping the client certificate from the connection. Ideally that would not be necessary.

SCEP was designed to work with plain HTTP (mainly). Client certificate support is even further from what's in the standard, considering HTTPS is mentioned as a (non ideal) option, but never mentioning mTLS at a point (as far as I know). The client authentication in SCEP does use a local certificate, but it's not working the same as mTLS. That said, it is interesting to see if this can be made to work.

> This is probably not possible, as client certificates are requested during TLS session establishment and the SCEP endpoint is only distinguishable once an HTTP request has been made.

Good point. It _may_ be an option to serve another TLS server that doesn't request a client certificate specifically for the SCEP provisioner, but that would change things up a bit.  

> macOS still doesn't save the CA certs provided in response, and as a result the client cert isn't actually usable by applications (or the system) unless the chain is installed to the keychain independently of the SCEP profile. This seems like a bug in macOS.

It may be a bug. It may also be by design. In SCEP a client should verify that the CA certificate returned in GetCACert is one that is expected. It's a bootstrap problem.

> If GetCACert responds with a certificate chain, the intermediate CA cert's keyUsage extension must include digitalSignature and keyEncipherment, or else the SCEP client will abort the request.

You're right about this. Due to how SCEP works, the CA needs to be able to both sign and (de|en)crypt messages. For this you'll currently need an RSA key in the intermediate. This is discussed [here](https://github.com/smallstep/certificates/pull/506#issuecomment-798987601). It's not easy to find, though, and internally the fact to document the SCEP support better has been brought up before. We can have another look at this.

I think this much of the above about mTLS and the CA root to trust is (at least partly) related to [7.10. Use of HTTP](https://www.rfc-editor.org/rfc/rfc8894.html#name-use-of-http). There's a difference between setting up the chain to the (trusted) root and what SCEP is doing with the GetCACert. This is what's discussed in the RFC.

That said, thank you for the detailed report and example environments. That'll help in getting this resolved.

**Comment by hslatman** [comment]
After a bit of searching I found that `profiles` is responsible for managing the provisioning files. `mdmclient` seems to be performing the SCEP requests; if not itself, then perhaps another library, but I haven't traced it.

I also found the Apple MDM Protocol Reference: https://developer.apple.com/business/documentation/MDM-Protocol-Reference.pdf. It mentions the following on page 209:

```
Initial Profiles Should Contain Only the Basics

The initial profile deployed to a device should contain only the following payloads:

• Any root certificates needed to establish SSL trust.
• Any intermediate certificates needed to establish SSL trust.
• A client identity certificate for use by the MDM payload (either a PKCS#12 container, or an SCEP payload). An SCEP payload is recommended.
• The MDM payload.
```

This seems to support the fact that the root (and intermediates) required to establish trust need to be provided in addition to a SCEP payload and be installed before doing further requests. `mdmclient` likely makes use of these (and perhaps also system default CA roots) when setting up an HTTPS connection to the SCEP CA. This has to be trusted, before any of the SCEP operations can take place. 

Then the SCEP protocol starts with GetCACert, which should at least return the certificate of the CA to which encrypted message should go. This CA does not need to be trusted, but the client can verify it (and MDM seems to support the fingerprint). This CA happens to also be the signer of the new certificate, because that's what currently makes most sense for `step-ca`. The SCEP client will be able to encrypt messages to the CA using its public key, will use a local self-signed certificate and can optionally provide a challenge password to authenticate itself.

The part I haven't checked/verified yet is the mTLS client certificate usage. Considering you're seeing the issue on renewals, it seems that `mdmclient` is in fact providing the previously minted certificate as a client certificate to `step-ca`. As mentioned before, I think mTLS is far from standard practice for SCEP. The MDM docs mention that MDM will only connect to secure MDM servers and that a device needs to authenticate itself using a client certificate, but the MDM server is _not_ the same as the SCEP server. It is interesting to see if it can be made to work though, as it seems to be quite a novel way to authenticate to a SCEP CA, but it'll likely require some changes.

**Comment by elyscape** [comment]
> I also found the Apple MDM Protocol Reference: https://developer.apple.com/business/documentation/MDM-Protocol-Reference.pdf.

That document is intended for vendors implementing MDM solutions, but `mdmclient` also handles configuration profiles deployed independently of an MDM enrollment. The reference for those, which I've been working off of, is [the Configuration Profile reference documentation][configprofile]. It's interesting that the MDM Protocol reference says that an initial profile should include "intermediate certificates needed to establish SSL trust", but that seems to be specifically for establishing trust of the HTTPS channel to the MDM server, and anyway if the server presents the intermediate certs then arguably they shouldn't be needed in the initial profile. Either way, SCEP payloads can be deployed as part of configuration profiles that aren't the initial profile, and in fact that's the scenario under which I ran into this issue.

> The MDM docs mention that MDM will only connect to secure MDM servers and that a device needs to authenticate itself using a client certificate, but the MDM server is _not_ the same as the SCEP server. It is interesting to see if it can be made to work though, as it seems to be quite a novel way to authenticate to a SCEP CA, but it'll likely require some changes.

To be clear, my use case isn't to use this for MDM and, even if it were, `mdmclient` still sends the SCEP challenge when it renews. This issue isn't a request for `step-ca` to accept mTLS as authentication for SCEP. The issue is that, when a client connects to `step-ca` with a client certificate but does not include the intermediate, the HTTPS listener cannot verify the chain and drops the connection. While this is poor behavior on the behalf of the client, `step-ca` should be able to handle it given that it issues certificates using the intermediate that would have signed the client certificate.

The second issue here, significantly less problematic, is that `step-ca` doesn't include the full CA cert chain in response to `GetCACert`, which some clients expect. That isn't as big a problem and fixing it would, in fact, cause additional issues, because of the third problem, which is that `step-ca` by default generates intermediate CA certificates that don't have all the required key usage flags set. This is a problem because apparently the macOS SCEP client checks for this and will refuse to request a certificate if they're missing, but only if `GetCACert` returns the full chain.

My recommendations for how to fix this issue are:

1. When `step-ca` is starting up, add the instance's active CA certificate (i.e. the intermediate it signs for) to [the `tls.Config` struct][tlsconfig] that is generated and used by [the HTTPS listener][listener].
2. Add a new certificate profile (maybe `scep-intermediate-ca`?) that is identical to [the `intermediate-ca` profile][intermediate], except that it also includes `digitalSignature` and `keyEncipherment` under `keyUsage`.
3. Update `GetCACert` to be able to return the full certificate chain of the active CA certificate rather than just the active CA cert.

In order to avoid breaking existing deployments that use the macOS SCEP client but have intermediate CA certs that are missing the proper key usage flags, item 2 should be called out in the release notes and item 3 should probably be gated on a flag in the SCEP provisioner config; the documentation should note that some clients expect it but that the intermediate CA cert needs to have been generated with the new profile for it to function properly.

[configprofile]: https://developer.apple.com/business/documentation/Configuration-Profile-Reference.pdf
[tlsconfig]: https://github.com/smallstep/certificates/blob/v0.18.0/ca/ca.go#L392-L443
[listener]: https://github.com/smallstep/certificates/blob/v0.18.0/ca/ca.go#L149
[intermediate]: https://github.com/smallstep/crypto/blob/ee37b94d634ed7b31b477d7535cad491446eeea9/x509util/templates.go#L149-L158

**Comment by dopey** [comment]
Hey @elyscape, we're still working at reproducing the issue. And, when I say "we", I mean @hslatman. We're kind of working on it between projects as we've got our hands full heading into the holidays. Progress may be a bit slow but we'll keep you updated. Cheers 🍻

**Comment by hslatman** [comment]
@elyscape: https://github.com/smallstep/certificates/pull/797 was just merged. The PR fixes the issue with mTLS and also allows configuration of returning the full CA chain and the encryption algorithm to use. It was tested successfully with (derivatives of) the CA config you provided. 

There's still room for improvement in terms of the configuration of the intermediate CA, like the `keyUsage` you mentioned. This might become part of a bigger work package related to SCEP. I'll see if we can at least improve the docs a bit about this before that.

**Comment by hslatman** [comment]
Some additional docs were merged with https://github.com/smallstep/docs/pull/105 and will soon be live. These cover the changes in https://github.com/smallstep/certificates/pull/797 (a.o.).
